### PR TITLE
Hedy 1.0.x: implement a way to switch the uplink-preset

### PR DESCRIPTION
**Pull request by SvenRoederer**
I started this idea on the Hedy-1.0.x branch and made it work to switch between the 3 current uplink-presets. Even I only tested for VPN03 --> tunnel-berlin and VPN03 --> no-tunnel it should also work for switching to VPN03-type.
For the tunnel-uplinks, the credentials need to be uploaded via luci-app, ffwizard or console, but it's working.

Any comments about the way I outlined here?
Some of these changes I already pushed to master, as they are valid in any case.

### CODE REVIEW

**Comment by pmelange** [comment]
I will try integrating the tunneldigger stuff.  Tunneldigger should also easily work in the 1.0.x firmwares (and 0.x.x too).

**Comment by SvenRoederer** [comment]
@pmelange So you feel this code acceptable, even I like to clean it a bit in the next step? Can the code be followed by an developer, who had not written it? Can be seen what's going on?

**Comment by SvenRoederer** [comment]
I rework the commits by squashing , rebased all to HEAD of "for-Hedy-1.0.x" branch.

@pmelange  please check that you tunneldigger-setup is right. should be very similar to uplink-notunnel

File: uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
+
 # always set correct masquerading, regardless of guard
 uci set firewall.zone_ffuplink.masq=1
 uci commit firewall
 
-# create ffberlin-uplink file an fill with basic settings
-uci -q get ffberlin-uplink || echo "" | uci import ffberlin-uplink
-uci >/dev/null -q get ffberlin-uplink.preset || uci set ffberlin-uplink.preset=settings
-uci >/dev/null -q get ffberlin-uplink.preset.current || uci set ffberlin-uplink.preset.current="no-tunnel"
 if [[ $(uci get ffberlin-uplink.preset.current) != "no-tunnel" ]]; then
```
Comment (pmelange) [inline uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel]: shouldn't this also be "notunnel"?

File: uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
-  uci rename ffberlin-uplink.preset.current=previous
+  # do not track preset when it was 'undefined', aka never configured
+  if [[ $(uci get ffberlin-uplink.preset.current) != 'undefined' ]]; then
+    logger -t "ffuplink" "uplink-preset has been changed."
+    uci rename ffberlin-uplink.preset.current=previous
+    create_ffuplink
+  fi
   uci set ffberlin-uplink.preset.current="no-tunnel"
```
Comment (pmelange) [inline uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel]: "notunnel" here too

**Comment by pmelange** [comment]
After this merge, we should also do #168

**Comment by SvenRoederer** [comment]
> The tunneldigger files are wrong. 

is this still relevant?

**Comment by pmelange** [comment]
I had PR #169 ready to merge with the needed tunneldigger changes, until this branch can force-pushed.  

Maybe you would like to try to cherry-pick the following:
f11a3a9c96557bfda33ae31875c9e9ecf8475cf5
631c5ae41eccf940f4063f5cd06730ef42ddc4c7

**Comment by SvenRoederer** [comment]
this rebase was required as if a conflicting commit to underlaying "for-Hedy-1.0.x" branch.

a rebase of PR #169 onto the updated "for-Hedy-1.0.x_change_uplink" should work w/o pain and fix it

**Comment by SvenRoederer** [comment]
but it not as expected. please check my comments from review of PR #169 and then we can pick ...

**Comment by pmelange** [comment]
OK. Still my rebasing is turning up all weird.  Here are the 3 commits.

2d06e56 
057fdb5 
cc67de1

File: uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel [hunk 3]
```diff
@@ -33,18 +37,12 @@ for byte in 2 3 4 5 6; do
...
 uci commit network.ffuplink_dev
 
-uci delete network.ffuplink
-uci set network.ffuplink=interface
-uci set network.ffuplink.ifname=ffuplink
+# add ffuplink_dev to br-wan if not there
+ifnames=$(uci get network.wan.ifname)
+list_contains ifnames ffuplink_wan || uci set network.wan.ifname="${ifnames} ffuplink_wan"
+uci commit network.wan
+
```
Comment (pmelange) [inline uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel]: There is no call to create_ffuplink.  Or if it's not needed, then a comment.

Comment (pmelange) [reply uplinks/freifunk-berlin-uplink-no-tunnel/uci-defaults/freifunk-berlin-z95_notunnel]: It's not easy to see that the call is in the if block above

**Comment by pmelange** [comment]
My comments don't seem to be working.  Here is the final list of commits to cherry-pick

 2d06e56
057fdb50
4507540

**Comment by pmelange** [comment]
I've gone ahead and done the cherry-pick and merging this branch.

**Comment by SvenRoederer** [comment]
I tested vpn03 --> tunnel-berlin with "Freifunk Berlin Hedy 1.0.2-SAm0815.uplink d40ceba" (using packages-feed 0e3b40f6d24bb085d2b8a0f2f38cd77e73c8e704 and had an incorrect setup.

faulty settings (at least):
* masquerading on ffuplink
* uci ffberlin-uplink.preset.*
* uci guard for the uplink-type

@pmelange  did you do a test before merging?

**Comment by pmelange** [comment]
I did not test vpn03.  I tested tunneldigger and default.  I went many times back and forth.  I didn't have any prolems there.  But take a look at the issues I have created 

https://github.com/freifunk-berlin/firmware/issues/636
https://github.com/freifunk-berlin/firmware/issues/637
https://github.com/freifunk-berlin/firmware/issues/638
https://github.com/freifunk-berlin/firmware/issues/639

This issue can also be added to the list.

**Comment by pmelange** [comment]
I have just added https://github.com/freifunk-berlin/firmware/issues/640

File: uplinks/freifunk-berlin-uplink-tunnelberlin-tunneldigger-files/uci-defaults/freifunk-berlin-z95_tunnelberlin-tunneldigger [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
+  uci set ffberlin-uplink.preset.current=${THIS_UPLINKNAME}
 fi
 # set set auth-type required for this uplink-type, e.g. for freifunk-wizard
 uci set ffberlin-uplink.uplink=settings
 uci set ffberlin-uplink.uplink.auth=none
+
 uci commit ffberlin-uplink
 
-. /lib/functions/guard.sh
+, /lib/functions/guard.sh
```
Comment (SvenRoederer) [inline uplinks/freifunk-berlin-uplink-tunnelberlin-tunneldigger-files/uci-defaults/freifunk-berlin-z95_tunnelberlin-tunneldigger]: this looks wrong, should be "." right?

Comment (pmelange) [reply uplinks/freifunk-berlin-uplink-tunnelberlin-tunneldigger-files/uci-defaults/freifunk-berlin-z95_tunnelberlin-tunneldigger]: You are right.  Commit adb6c3d0c295994acaf0b705358b8db728e54b48 has been added to for-Hedy-1.0.x_change_uplink

**Comment by SvenRoederer** [comment]
the problems mentioned in https://github.com/freifunk-berlin/firmware-packages/pull/159#issuecomment-445949509 are gone, seems that I was to much in a hurry

**Comment by SvenRoederer** [comment]
> I tested vpn03 --> tunnel-berlin with "Freifunk Berlin Hedy 1.0.2-SAm0815.uplink d40ceba" (using packages-feed [0e3b40f](https://github.com/freifunk-berlin/firmware-packages/commit/0e3b40f6d24bb085d2b8a0f2f38cd77e73c8e704) and had an incorrect setup.
> 
> faulty settings (at least):
> 
>     * masquerading on ffuplink
> 
>     * uci ffberlin-uplink.preset.*
> 
>     * uci guard for the uplink-type
> 
> 
> @pmelange did you do a test before merging?

the problems mentioned in https://github.com/freifunk-berlin/firmware-packages/pull/159#issuecomment-445949509 are gone, seems that I was to much in a hurry

**Comment by pmelange** [comment]
Good. Merge the changes and let's test some more

**Pull request merged**